File: src/main/java/org/danilopianini/util/FlexibleQuadTree.java
/*******************************************************************************
 * Copyright (C) 2009-2017, Danilo Pianini and contributors
 * listed in the project's build.gradle or pom.xml file.
 *
 * This file is distributed under the terms of the Apache License, version 2.0
 *******************************************************************************/
package org.danilopianini.util;

import com.google.common.hash.Hashing;

import javax.annotation.Nonnull;
import java.io.Serializable;
import java.util.ArrayDeque;
import java.util.ArrayList;
import java.util.Deque;
import java.util.EnumMap;
import java.util.Iterator;
import java.util.List;
import java.util.Map;
import java.util.Objects;

import static java.lang.Math.ceil;
import static java.lang.Math.floor;
import static java.lang.Math.max;
import static java.lang.Math.min;
import static java.lang.Math.nextDown;
import static java.lang.Math.nextUp;

/**
 * @param <E> the type of objects stored in the quadtree
 */
public final class FlexibleQuadTree<E> implements SpatialIndex<E> {

    /**
     * Default maximum number of entries per node.
     */
    public static final int DEFAULT_CAPACITY = 10;
    private static final long serialVersionUID = 0L;
    private final Map<Child, FlexibleQuadTree<E>> children;
    private final Deque<QuadTreeEntry<E>> elements;
    private final int maxElements;
    private Rectangle2D bounds;
    private FlexibleQuadTree<E> parent;

    /**
     * root is NOT consistent everywhere. It is only guaranteed to be consistent
     * in the entry point node and in the current root.
     */
    private FlexibleQuadTree<E> root;

    /**
     * Builds a {@link FlexibleQuadTree} with the default node capacity.
     */
    public FlexibleQuadTree() {
        this(DEFAULT_CAPACITY);
    }

    private FlexibleQuadTree(
        final double minx,
        final double maxx,
        final double miny,
        final double maxy,
        final int elemPerQuad,
        final FlexibleQuadTree<E> rootNode,
        final FlexibleQuadTree<E> parentNode
    ) {
        this(elemPerQuad, rootNode, parentNode);
        bounds = new Rectangle2D(minx, miny, maxx, maxy);
    }

    private FlexibleQuadTree(
        final int elemPerQuad,
        final FlexibleQuadTree<E> rootNode,
        final FlexibleQuadTree<E> parentNode
    ) {
        if (elemPerQuad < 2) {
            throw new IllegalArgumentException("At least two elements per quadtree are required for this index to work properly");
        }
        elements = new ArrayDeque<>(DEFAULT_CAPACITY);
        children = new EnumMap<>(Child.class);
        maxElements = elemPerQuad;
        parent = parentNode;
        root = rootNode == null ? this : rootNode;
    }

    /**
     * @param elemPerQuad maximum number of elements per quad
     */
    public FlexibleQuadTree(final int elemPerQuad) {
        this(elemPerQuad, null, null);
    }

    private double centerX() {
        return bounds.getCenterX();
    }

    private double centerY() {
        return bounds.getCenterY();
    }

    private boolean contains(final double x, final double y) {
        return bounds == null || bounds.contains(x, y);
    }

    private FlexibleQuadTree<E> create(
        final double minx,
        final double maxx,
        final double miny,
        final double maxy,
        final FlexibleQuadTree<E> father
    ) {
        return new FlexibleQuadTree<>(minx, maxx, miny, maxy, getMaxElementsNumber(), root, father);
    }

    private void createChildIfAbsent(final Child c) {
        children.putIfAbsent(c, create(minX(c), maxX(c), minY(c), maxY(c), this));
    }

    private void createParent(final double x, final double y) {
        /*
         * Determine where the parent should be
         */
        if (x < centerX()) {
            final double minx = 2 * minX() - maxX();
            if (y < centerY()) {
                /*
                 * This will be TR child of the new parent
                 */
                root = create(minx, maxX(), 2 * minY() - maxY(), maxY(), null);
                root.setChild(Child.TR, this);
            } else {
                /*
                 * This will be BR child of the new parent
                 */
                root = create(minx, maxX(), minY(), 2 * maxY() - minY(), null);
                root.setChild(Child.BR, this);
            }
        } else {
            final double maxx = 2 * maxX() - minX();
            if (y < centerY()) {
                /*
                 * This will be TL child of the new parent
                 */
                root = create(minX(), maxx, 2 * minY() - maxY(), maxY(), null);
                root.setChild(Child.TL, this);
            } else {
                /*
                 * This will be BL child of the new parent
                 */
                root = create(minX(), maxx, minY(), 2 * maxY() - minY(), null);
                root.setChild(Child.BL, this);
            }
        }
        /*
         * A bit cryptic, but the root of the new root is the root itself.
         * Otherwise, the root would point to the previous root.
         */
        root.root = root;
        root.subdivide();
    }

    @Override
    public int getDimensions() {
        return 2;
    }

    /**
     * @return the maximum number of elements per node
     */
    public int getMaxElementsNumber() {
        return maxElements;
    }

    private boolean hasSpace() {
        return elements.size() < maxElements;
    }

    @Override
    public void insert(final E e, final double... pos) {
        assert pos.length == 2;
        insert(e, pos[0], pos[1]);
    }

    /**
     * Same of {@link #insert(Object, double...)}, but with explicit parameters.
     *
     * @param e element
     * @param x X
     * @param y Y
     */
    public void insert(final E e, final double x, final double y) {
        if (bounds == null) {
            if (hasSpace()) {
                insertNode(e, x, y);
                return;
            }
            double minx = Double.POSITIVE_INFINITY;
            double miny = Double.POSITIVE_INFINITY;
            double maxx = Double.NEGATIVE_INFINITY;
            double maxy = Double.NEGATIVE_INFINITY;
            for (final QuadTreeEntry<E> element : elements) {
                minx = min(minx, element.x);
                miny = min(miny, element.y);
                maxx = max(maxx, element.x);
                maxy = max(maxy, element.y);
            }
            assert Double.isFinite(minx);
            assert Double.isFinite(maxx);
            assert Double.isFinite(miny);
            assert Double.isFinite(maxy);
            bounds = new Rectangle2D(floor(nextDown(minx)), floor(nextDown(miny)), ceil(nextUp(maxx)), ceil(nextUp(maxy)));
        }
        /*
         * I must insert starting from the root. If the root does not contain
         * the coordinates, then the tree should be expanded upwards
         */
        while (!root.contains(x, y)) {
            root.createParent(x, y);
            root = root.root;
        }
        root.insertHere(e, x, y);
    }

    private void insertHere(final E e, final double x, final double y) {
        if (hasSpace()) {
            insertNode(e, x, y);
        } else {
            if (children.isEmpty()) {
                subdivide();
            }
            selectChild(x, y).insertHere(e, x, y);
        }
    }

    private void insertNode(@Nonnull final E e, final double x, final double y) {
        assert elements.size() < maxElements : "Bug in " + getClass() + ". Forced insertion over the container size.";
        elements.push(new QuadTreeEntry<>(e, x, y));
    }

    private double maxX() {
        return bounds.getMaxX();
    }

    private double maxX(final Child c) {
        switch (c) {
            case TR:
            case BR:
                return maxX();
            case BL:
            case TL:
                return centerX();
            default:
                throw new IllegalStateException();
        }
    }

    private double maxY() {
        return bounds.getMaxY();
    }

    private double maxY(final Child c) {
        switch (c) {
            case BL:
            case BR:
                return centerY();
            case TR:
            case TL:
                return maxY();
            default:
                throw new IllegalStateException();
        }
    }

    private double minX() {
        return bounds.getMinX();
    }

    private double minX(final Child c) {
        switch (c) {
            case TR:
            case BR:
                return centerX();
            case BL:
            case TL:
                return minX();
            default:
                throw new IllegalStateException();
        }
    }

    private double minY() {
        return bounds.getMinY();
    }

    private double minY(final Child c) {
        switch (c) {
            case BL:
            case BR:
                return minY();
            case TR:
            case TL:
                return centerY();
            default:
                throw new IllegalStateException();
        }
    }

    /**
     * Same of {@link #move(Object, double[], double[])}, but with explicit
     * parameters.
     *
     * @param e  the element
     * @param sx the start x
     * @param sy the start y
     * @param fx the final x
     * @param fy the final y
     *
     * @return true if the element is found and no error occurred
     */
    public boolean move(final E e, final double sx, final double sy, final double fx, final double fy) {
        return moveFromNode(root, e, sx, sy, fx, fy);
    }

    @Override
    public boolean move(final E e, final double[] start, final double[] end) {
        assert start.length == 2;
        assert end.length == 2;
        return move(e, start[0], start[1], end[0], end[1]);
    }

    private boolean moveFromNode(
            final FlexibleQuadTree<E> root,
            final E e,
            final double sx, final double sy,
            final double fx, final double fy) {
        final QuadTreeEntry<E> toRemove = new QuadTreeEntry<>(e, sx, sy);
        for (FlexibleQuadTree<E> cur = root; cur.contains(sx, sy); cur = cur.selectChild(sx, sy)) {
            if (cur.elements.remove(toRemove)) {
                /*
                 * Node found.
                 */
                if (cur.contains(fx, fy)) {
                    /*
                     * Moved within the same quadrant.
                     */
                    cur.insertNode(e, fx, fy);
                } else if (
                    cur.parent == null
                    || !cur.parent.contains(fx, fy)
                    || !cur.swapMostStatic(e, fx, fy)
                ) {
                    /*
                     * In case:
                     *  - we are the root
                     *  - we moved outside the parent's area
                     *  - the swapping operation failed
                     */
                    insert(e, fx, fy);
                }
                return true;
            }
            if (cur.children.isEmpty()) {
                return false;
            }
        }
        return false;
    }

    /**
     * Same of querying with arrays, but with explicit parameters.
     *
     * @param x1 Rectangle X coordinate of the first point
     * @param y1 Rectangle Y coordinate of the first point
     * @param x2 Rectangle X coordinate of the second point
     * @param y2 Rectangle Y coordinate of the second point
     *
     * @return {@link List} of Objects in range.
     */
    public List<E> query(final double x1, final double y1, final double x2, final double y2) {
        final List<E> result = new ArrayList<>();
        root.query(min(x1, x2), min(y1, y2), max(x1, x2), max(y1, y2), result);
        return result;
    }

    private void query(// NOPMD: False positive
        final double sx,
        final double sy,
        final double fx,
        final double fy,
        final List<E> results
    ) {
        assert !(bounds == null && !children.isEmpty());
        if (bounds == null || bounds.intersects(sx, sy, fx, fy)) {
            for (final QuadTreeEntry<E> entry : elements) {
                if (entry.isIn(sx, sy, fx, fy)) {
                    results.add(entry.element);
                }
            }
            // If there are no children, this will skip them.
            for (final FlexibleQuadTree<E> childOpt : children.values()) {
                childOpt.query(sx, sy, fx, fy, results);
            }
        }
    }

    @Override
    public List<E> query(final double[]... space) {
        if (space.length != 2 || space[0].length != 2 || space[1].length != 2) {
            throw new IllegalArgumentException();
        }
        return query(space[0][0], space[0][1], space[1][0], space[1][1]);
    }

    @Override
    public boolean remove(final E e, final double... pos) {
        assert pos.length == 2;
        return remove(e, pos[0], pos[1]);
    }

    /**
     * Same of {@link #remove(Object, double...)} with explicit parameters.
     *
     * @param e Element to remove
     * @param x X position of the element
     * @param y Y position of the element
     *
     * @return true if the element has been found and removed
     */
    public boolean remove(final E e, final double x, final double y) {
        return root.removeHere(e, x, y);
    }

    private boolean removeHere(final E e, final double x, final double y) {
        if (contains(x, y)) {
            return elements.remove(new QuadTreeEntry<>(e, x, y)) || removeInChildren(e, x, y);
        }
        return false;
    }

    private boolean removeInChildren(final E e, final double x, final double y) {
        return children.values().stream().anyMatch(c -> c.removeHere(e, x, y));
    }

    private FlexibleQuadTree<E> selectChild(final double x, final double y) {
        assert !children.isEmpty();
        if (x < centerX()) {
            return y < centerY() ? children.get(Child.BL) : children.get(Child.TL);
        } else {
            return y < centerY() ? children.get(Child.BR) : children.get(Child.TR);
        }
    }

    private void setChild(final Child c, final FlexibleQuadTree<E> child) {
        if (children.put(c, child) != null) {
            throw new IllegalStateException();
        }
        child.parent = this;
    }

    private void subdivide() {
        for (final Child c : Child.values()) {
            createChildIfAbsent(c);
        }
    }

    private boolean swapMostStatic(final E e, final double fx, final double fy) {
        assert parent != null : "Tried to swap on a null parent.";
        final Iterator<QuadTreeEntry<E>> iterator = parent.elements.descendingIterator();
        while (iterator.hasNext()) {
            final QuadTreeEntry<E> target = iterator.next();
            if (contains(target.x, target.y)) {
                /*
                 * There is a swappable node
                 */
                iterator.remove();
                elements.push(target);
                parent.insertNode(e, fx, fy);
                return true;
            }
        }
        return false;
    }

    @Override
    public String toString() {
        return (bounds == null ? "Unbounded" : bounds.toString()) + ":" + elements.toString();
    }

    private enum Child {
        TR, BR, BL, TL
    }

    private static class QuadTreeEntry<E> implements Serializable {
        private static final long serialVersionUID = 9021533648086596986L;
        private final E element;
        private final double x, y;

        QuadTreeEntry(final E el, final double xp, final double yp) {
            element = el;
            x = xp;
            y = yp;
        }

        @Override
        public boolean equals(final Object obj) {
            if (obj == this) {
                return true;
            }
            if (obj instanceof QuadTreeEntry<?>) {
                final QuadTreeEntry<?> e = (QuadTreeEntry<?>) obj;
                // NOPMD: Pointer comparison intended
                return samePosition(e) && Objects.equals(element, e.element);
            }
            return false;
        }

        @Override
        @SuppressWarnings("UnstableApiUsage")
        public int hashCode() {
            return Hashing.murmur3_32_fixed().newHasher()
                .putDouble(x)
                .putDouble(y)
                .putInt(element.hashCode())
                .hash()
                .asInt();
        }

        public boolean isIn(final double sx, final double sy, final double fx, final double fy) {
            return x >= sx && x < fx && y >= sy && y < fy;
        }

        public boolean samePosition(final QuadTreeEntry<?> target) {
            return x == target.x && y == target.y;
        }

        @Override
        public String toString() {
            return element.toString() + "@[" + x + ", " + y + "]";
        }
    }

    private static class Rectangle2D implements Serializable {
        private static final long serialVersionUID = -7890062202005580979L;
        private final double minx, miny, maxx, maxy;

        Rectangle2D(final double sx, final double sy, final double fx, final double fy) {
            minx = min(sx, fx);
            miny = min(sy, fy);
            maxx = max(sx, fx);
            maxy = max(sy, fy);
        }

        public boolean contains(final double x, final double y) {
            return x >= minx && y >= miny && x < maxx && y < maxy;
        }

        public double getCenterX() {
            return minx + (maxx - minx) / 2;
        }

        public double getCenterY() {
            return miny + (maxy - miny) / 2;
        }

        public double getMaxX() {
            return maxx;
        }

        public double getMaxY() {
            return maxy;
        }

        public double getMinX() {
            return minx;
        }

        public double getMinY() {
            return miny;
        }

        public boolean intersects(final double sx, final double sy, final double fx, final double fy) {
            return fx >= minx && fy >= miny && sx < maxx && sy < maxy;
        }

        @Override
        public String toString() {
            return "[" + minx + "," + miny + " - " + maxx + "," + maxy + "]";
        }
    }
}

File: src/main/kotlin/org/danilopianini/util/FlexibleQuadTree.kt
/*
 * Copyright (C) 2009-2017, Danilo Pianini and contributors
 * listed in the project's build.gradle or pom.xml file.
 * This file is distributed under the terms of the Apache License, version 2.0
 */
package org.danilopianini.util

import java.io.Serializable
import java.util.ArrayDeque
import java.util.Deque
import java.util.EnumMap
import kotlin.math.ceil
import kotlin.math.floor
import kotlin.math.max
import kotlin.math.min
import kotlin.math.nextDown
import kotlin.math.nextUp

/**
 * @param <E> the type of objects stored in the quadtree
 */
class FlexibleQuadTree<E> private constructor(
    elementsPerQuad: Int,
    rootNode: FlexibleQuadTree<E>?,
    parentNode: FlexibleQuadTree<E>?,
) : SpatialIndex<E> {
    private val children: MutableMap<Child, FlexibleQuadTree<E>>
    private val elements: Deque<QuadTreeEntry<E>>

    /**
     * @return the maximum number of elements per node
     */
    val maxElementsNumber: Int
    private var bounds: Rectangle2D? = null
    private var parent: FlexibleQuadTree<E>?

    /**
     * root is NOT consistent everywhere. It is only guaranteed to be consistent
     * in the entry point node and in the current root.
     */
    private var root: FlexibleQuadTree<E>

    private constructor(
        minX: Double,
        maxX: Double,
        minY: Double,
        maxY: Double,
        elemPerQuad: Int,
        rootNode: FlexibleQuadTree<E>?,
        parentNode: FlexibleQuadTree<E>?,
    ) : this(elemPerQuad, rootNode, parentNode) {
        bounds = Rectangle2D(minX, minY, maxX, maxY)
    }

    init {
        require(elementsPerQuad >= 2) {
            "At least two elements per quadtree are required for this index to work properly"
        }
        elements = ArrayDeque<QuadTreeEntry<E>>(DEFAULT_CAPACITY)
        children = EnumMap<Child, FlexibleQuadTree<E>>(Child::class.java)
        this.maxElementsNumber = elementsPerQuad
        parent = parentNode
        root = rootNode ?: this
    }

    /**
     * Builds a [FlexibleQuadTree] with the default node capacity.
     * @param elemPerQuad maximum number of elements per quad
     */
    @JvmOverloads
    constructor(elemPerQuad: Int = DEFAULT_CAPACITY) : this(elemPerQuad, null, null)

    private val centerX: Double get() = checkNotNull(bounds).centerX

    private val centerY: Double get() = checkNotNull(bounds).centerY

    private fun contains(
        x: Double,
        y: Double,
    ): Boolean = bounds == null || checkNotNull(bounds).contains(x, y)

    private fun create(
        minX: Double,
        maxX: Double,
        minY: Double,
        maxY: Double,
        father: FlexibleQuadTree<E>?,
    ): FlexibleQuadTree<E> = FlexibleQuadTree<E>(minX, maxX, minY, maxY, this.maxElementsNumber, root, father)

    private fun createChildIfAbsent(c: Child) {
        children.putIfAbsent(c, create(minX(c), maxX(c), minY(c), maxY(c), this))
    }

    private fun createParent(
        x: Double,
        y: Double,
    ) {
        /*
         * Determine where the parent should be
         */
        if (x < centerX) {
            val minx = 2 * minX - maxX
            if (y < centerY) {
                /*
                 * This will be the TR child of the new parent
                 */
                root = create(minx, maxX, 2 * minY - maxY, maxY, null)
                root.setChild(Child.TR, this)
            } else {
                /*
                 * This will be the BR child of the new parent
                 */
                root = create(minx, maxX, minY, 2 * maxY - minY, null)
                root.setChild(Child.BR, this)
            }
        } else {
            val maxX = 2 * maxX - minX
            if (y < centerY) {
                /*
                 * This will be the TL child of the new parent
                 */
                root = create(minX, maxX, 2 * minY - maxY, maxY, null)
                root.setChild(Child.TL, this)
            } else {
                /*
                 * This will be the BL child of the new parent
                 */
                root = create(minX, maxX, minY, 2 * maxY - minY, null)
                root.setChild(Child.BL, this)
            }
        }
        /*
         * A bit cryptic, but the root of the new root is the root itself.
         * Otherwise, the root would point to the previous root.
         */
        root.root = root
        root.subdivide()
    }

    override val dimensions: Int = 2

    private fun hasSpace(): Boolean = elements.size < this.maxElementsNumber

    override fun insert(
        element: E,
        vararg pos: Double,
    ) {
        assert(pos.size == 2)
        insert(element, pos[0], pos[1])
    }

    /**
     * Same of [.insert], but with explicit parameters.
     *
     * @param element element
     * @param x X
     * @param y Y
     */
    fun insert(
        element: E,
        x: Double,
        y: Double,
    ) {
        if (bounds == null) {
            if (hasSpace()) {
                insertNode(element, x, y)
                return
            }
            var minX = Double.Companion.POSITIVE_INFINITY
            var minY = Double.Companion.POSITIVE_INFINITY
            var maxX = Double.Companion.NEGATIVE_INFINITY
            var maxY = Double.Companion.NEGATIVE_INFINITY
            for (currentElement in elements) {
                minX = min(minX, currentElement.x)
                minY = min(minY, currentElement.y)
                maxX = max(maxX, currentElement.x)
                maxY = max(maxY, currentElement.y)
            }
            assert(java.lang.Double.isFinite(minX))
            assert(java.lang.Double.isFinite(maxX))
            assert(java.lang.Double.isFinite(minY))
            assert(java.lang.Double.isFinite(maxY))
            bounds =
                Rectangle2D(floor(minX.nextDown()), floor(minY.nextDown()), ceil(maxX.nextUp()), ceil(maxY.nextUp()))
        }
        /*
         * I must insert starting from the root. If the root does not contain
         * the coordinates, then the tree should be expanded upwards
         */
        while (!root.contains(x, y)) {
            root.createParent(x, y)
            root = root.root
        }
        root.insertHere(element, x, y)
    }

    private fun insertHere(
        e: E,
        x: Double,
        y: Double,
    ) {
        if (hasSpace()) {
            insertNode(e, x, y)
        } else {
            if (children.isEmpty()) {
                subdivide()
            }
            selectChild(x, y).insertHere(e, x, y)
        }
    }

    private fun insertNode(
        e: E,
        x: Double,
        y: Double,
    ) {
        assert(elements.size < this.maxElementsNumber) {
            "Bug in $javaClass. Forced insertion over the container size."
        }
        elements.push(QuadTreeEntry<E>(e, x, y))
    }

    private val maxX: Double get() = checkNotNull(bounds).maxX

    private fun maxX(c: Child): Double =
        when (c) {
            Child.TR, Child.BR -> maxX
            Child.BL, Child.TL -> centerX
        }

    private val maxY: Double get() = checkNotNull(bounds).maxY

    private fun maxY(c: Child): Double =
        when (c) {
            Child.BL, Child.BR -> centerY
            Child.TR, Child.TL -> maxY
        }

    private val minX: Double get() = checkNotNull(bounds).minX

    private fun minX(c: Child): Double =
        when (c) {
            Child.TR, Child.BR -> centerX
            Child.BL, Child.TL -> minX
        }

    private val minY: Double get() = checkNotNull(bounds).minY

    private fun minY(c: Child): Double =
        when (c) {
            Child.BL, Child.BR -> minY
            Child.TR, Child.TL -> centerY
        }

    /**
     * Same of [.move], but with explicit
     * parameters.
     *
     * @param e  the element
     * @param sx the start x
     * @param sy the start y
     * @param fx the final x
     * @param fy the final y
     *
     * @return true if the element is found and no error occurred
     */
    fun move(
        e: E,
        sx: Double,
        sy: Double,
        fx: Double,
        fy: Double,
    ): Boolean {
        val toRemove = QuadTreeEntry<E>(e, sx, sy)
        var cur = root
        var moved = false
        while (cur.contains(sx, sy) && !moved) {
            when {
                cur.elements.remove(toRemove) -> {
                    /*
                     * Node found.
                     */
                    moved = true
                    val currentParent = cur.parent
                    if (cur.contains(fx, fy)) {
                        /*
                         * Moved within the same quadrant.
                         */
                        cur.insertNode(e, fx, fy)
                    } else if (
                        // We are root, or
                        currentParent == null ||
                        // we moved outside the parent's area, or
                        !currentParent.contains(fx, fy) ||
                        // the swapping operation failed
                        !cur.swapMostStatic(e, fx, fy)
                    ) {
                        insert(e, fx, fy)
                    }
                }
                cur.children.isEmpty() -> break
                else -> cur = cur.selectChild(sx, sy)
            }
        }
        return moved
    }

    override fun move(
        e: E,
        start: DoubleArray,
        end: DoubleArray,
    ): Boolean {
        assert(start.size == 2)
        assert(end.size == 2)
        return move(e, start[0], start[1], end[0], end[1])
    }

    /**
     * Same of querying with arrays, but with explicit parameters.
     *
     * @param x1 Rectangle X coordinate of the first point
     * @param y1 Rectangle Y coordinate of the first point
     * @param x2 Rectangle X coordinate of the second point
     * @param y2 Rectangle Y coordinate of the second point
     *
     * @return [List] of Objects in range.
     */
    fun query(
        x1: Double,
        y1: Double,
        x2: Double,
        y2: Double,
    ): MutableList<E> {
        val result: MutableList<E> = ArrayList<E>()
        root.query(min(x1, x2), min(y1, y2), max(x1, x2), max(y1, y2), result)
        return result
    }

    private fun query(
        sx: Double,
        sy: Double,
        fx: Double,
        fy: Double,
        results: MutableList<E>,
    ) {
        assert(!(bounds == null && !children.isEmpty()))
        if (bounds == null || checkNotNull(bounds).intersects(sx, sy, fx, fy)) {
            for (entry in elements) {
                if (entry.isIn(sx, sy, fx, fy)) {
                    results.add(entry.element)
                }
            }
            // If there are no children, this will skip them.
            for (childOpt in children.values) {
                childOpt.query(sx, sy, fx, fy, results)
            }
        }
    }

    override fun query(vararg space: DoubleArray): MutableList<E> {
        require(space.size == 2 && space[0].size == 2 && space[1].size == 2) {
            "Invalid space definition"
        }
        return query(space[0][0], space[0][1], space[1][0], space[1][1])
    }

    override fun remove(
        e: E,
        vararg pos: Double,
    ): Boolean {
        assert(pos.size == 2)
        return remove(e, pos[0], pos[1])
    }

    /**
     * Same of [.remove] with explicit parameters.
     *
     * @param e Element to remove
     * @param x X position of the element
     * @param y Y position of the element
     *
     * @return true if the element has been found and removed
     */
    fun remove(
        e: E,
        x: Double,
        y: Double,
    ): Boolean = root.removeHere(e, x, y)

    private fun removeHere(
        e: E,
        x: Double,
        y: Double,
    ): Boolean {
        if (contains(x, y)) {
            return elements.remove(QuadTreeEntry<E>(e, x, y)) || removeInChildren(e, x, y)
        }
        return false
    }

    private fun removeInChildren(
        e: E,
        x: Double,
        y: Double,
    ): Boolean = children.values.any { it.removeHere(e, x, y) }

    private fun selectChild(
        x: Double,
        y: Double,
    ): FlexibleQuadTree<E> {
        require(children.isNotEmpty())
        val result =
            when {
                x < centerX && y < centerY -> children[Child.BL]
                x < centerX && y >= centerY -> children[Child.TL]
                x >= centerX && y < centerY -> children[Child.BR]
                else -> children[Child.TR]
            }
        return checkNotNull(result) {
            "Child node at ($x, $y) is null. " +
                "This indicates an inconsistency in the quadtree structure, as all children should be initialized."
        }
    }

    private fun setChild(
        c: Child,
        child: FlexibleQuadTree<E>,
    ) {
        check(children.put(c, child) == null)
        child.parent = this
    }

    private fun subdivide() {
        for (c in Child.entries) {
            createChildIfAbsent(c)
        }
    }

    private fun swapMostStatic(
        e: E,
        fx: Double,
        fy: Double,
    ): Boolean {
        val myParent = checkNotNull(parent) { "Tried to swap on a null parent." }
        val iterator = myParent.elements.descendingIterator()
        while (iterator.hasNext()) {
            val target = iterator.next()
            if (contains(target.x, target.y)) {
                /*
                 * There is a swappable node
                 */
                iterator.remove()
                elements.push(target)
                myParent.insertNode(e, fx, fy)
                return true
            }
        }
        return false
    }

    override fun toString(): String = (bounds?.toString() ?: "Unbounded") + ":" + elements.toString()

    private enum class Child {
        TR,
        BR,
        BL,
        TL,
    }

    private data class QuadTreeEntry<E>(
        val element: E,
        val x: Double,
        val y: Double,
    ) : Serializable {
        fun isIn(
            sx: Double,
            sy: Double,
            fx: Double,
            fy: Double,
        ): Boolean = x >= sx && x < fx && y >= sy && y < fy

        override fun toString(): String = "$element@[$x, $y]"

        companion object {
            private const val serialVersionUID = 1L
        }
    }

    private class Rectangle2D(
        sx: Double,
        sy: Double,
        fx: Double,
        fy: Double,
    ) : Serializable {
        val minX: Double = min(sx, fx)
        val minY: Double = min(sy, fy)
        val maxX: Double = max(sx, fx)
        val maxY: Double = max(sy, fy)

        fun contains(
            x: Double,
            y: Double,
        ): Boolean = x >= this.minX && y >= this.minY && x < this.maxX && y < this.maxY

        val centerX: Double
            get() = this.minX + (this.maxX - this.minX) / 2

        val centerY: Double
            get() = this.minY + (this.maxY - this.minY) / 2

        fun intersects(
            sx: Double,
            sy: Double,
            fx: Double,
            fy: Double,
        ): Boolean = fx >= this.minX && fy >= this.minY && sx < this.maxX && sy < this.maxY

        override fun toString(): String = "[" + this.minX + "," + this.minY + " - " + this.maxX + "," + this.maxY + "]"

        companion object {
            private const val serialVersionUID = 1L
        }
    }

    /**
     * Constants.
     */
    companion object {
        /**
         * Default maximum number of entries per node.
         */
        const val DEFAULT_CAPACITY: Int = 10
    }
}
